lib/ws_servers/api/algos/algo_worker.js

```diff
@@ -0,0 +1,188 @@
+    const instances = this.host.getAOInstances()
+    const activeInstances = instances.filter((aoInstance) => {
+      const { state = {} } = aoInstance
+      const { active } = state
+      return active
+    })
+
+    const algos = activeInstances.map((aoInstance) => {
+      const { state = {} } = aoInstance
+      const { gid, name, args, label } = state
+
+      return [gid, name, label, args]
+    })
+
+    this.pub(['data.aos', 'bitfinex', algos])
+  }
+
+  close () {
+    if (!this.host) {
+      return
+    }
+
+    this.host.removeAllListeners()
+    this.host.close()
+    this.host.cleanState()
+  }
+
+  reconnect (dms) {
+    if (!this.host) {
+      return
+    }
+
+    const adapter = this.host.getAdapter()
+
+    if (adapter.updateAuthArgs) {
+      adapter.updateAuthArgs({ dms: dms ? 4 : 0 })
+    }
+
+    this.host.reconnect()
+    d('issued reconnect [dms %s]', dms)
+  }
+
+  registerEvents () {
+    this.host.on('error', (error) => {
+      this.sendError(`error: ${error}`)
+    })
+
+    this.host.on('ao:start', (instance) => {
+      const { state = {} } = instance
+      const { name, label, args, gid } = state
+
+      d('ao started: %s %s', name, label)
+
+      this.sendSuccess(`Started AO ${name} on Bitfinex`)
+      this.pub(['data.ao', 'bitfinex', { gid, name, label, args }])
+    })
+
+    this.host.on('ao:stop', (instance) => {
+      const { state = {} } = instance
+      const { gid } = state
+
+      d('ao stopped: %s', gid)
+
+      this.sendSuccess('Stopped algo order')
+      this.pub(['data.ao.stopped', 'bitfinex', gid])
+    })
+
+    this.host.on('ao:persist:db:update', async (updateOpts) => {
+      await this._updateAlgo(updateOpts)
+    })
+
+    this.host.on('meta:reload', async () => {
+      d('meta reloaded')
+      this.pub(['algo.reload'])
+    })
+
+    this.host.on('ao:loaded', (gid) => {
+      d('ao loaded: %s', gid)
+      this.pub(['algo.order_loaded', gid])
+    })
+  }
+
+  async _updateAlgo (updateOpts) {
+    const { AlgoOrder } = this.algoDB
+    await AlgoOrder.set(updateOpts)
+    d('ao instance updated %s', updateOpts.gid)
+  }
+
+  sendSuccess (msg) {
+    this.pub(['notify', 'success', msg])
+  }
+
+  sendError (msg) {
+    this.pub(['notify', 'error', msg])
+  }
+
+  async cancelOrder (gid) {
+    const { host } = this
+
+    if (!host.getAOInstance(gid)) {
+      throw new Error('Requested algo order not running, cannot stop')
+    }
+
+    await host.stopAO(gid)
+    d('stopped AO for user %s on gid: %s', this.userID, gid)
+  }
+
+  async submitOrder (aoID, order) {
+    const { host } = this
+
+    const validationError = validateAO(host, aoID, order)
+
+    if (validationError) {
+      throw new Error(validationError)
+    }
+
+    try {
+      const gid = await host.startAO(aoID, order)
+      d('started AO for user %s on [%s]', this.userID, gid)
+    } catch (e) {
+      d('error starting AO %s: %s for %s', aoID, e, this.userID)
+      console.error(aoID, e.stack)
```

Comment:
Will `console.error` stay or was it for dev debugging purposes only?

Reply:
let me double check but i think it should stay. original line of code was 

```js
d('error starting AO %s for %s: %s', aoID, exID, e.stack)
```

and i think the stack prop was not rendered nicely

Reply:
its rendered in the same way, readding...